### PR TITLE
Fix haddocks generation - now using nix

### PR DESCRIPTION
**Pull request by carbolymer**
# Changelog

```yaml
- description: |
    Fix haddocks generation - now using nix
# uncomment types applicable to the change:
  type:
  # - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
   - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
   - documentation  # change in code docs, haddocks...
```

# Context

Fixes the same issue which happened in cardano-api repo: 
- https://github.com/IntersectMBO/cardano-api/issues/437

This PR copies changes from cardano-api from:
- https://github.com/IntersectMBO/cardano-api/pull/363
- https://github.com/IntersectMBO/cardano-api/pull/577

Sample workflow run with generated haddocks (attached as an artifact): https://github.com/IntersectMBO/cardano-cli/actions/runs/9796004207

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

### CODE REVIEW

**Comment by carbolymer** [comment]
FIXME:
* [x] generated haddocks contain only docs for the tests - docs for library and exe are missing

**Comment by smelc** [comment]
@andreabedini> would you have a clue why this PR only produces haddocks for tests (see @carbolymer's comment above: https://github.com/input-output-hk/cardano-cli/pull/445#issuecomment-1820875560)?

**Comment by andreabedini** [comment]
> @andreabedini> would you have a clue why this PR only produces haddocks for tests (see @carbolymer's comment above: [#445 (comment)](https://github.com/input-output-hk/cardano-cli/pull/445#issuecomment-1820875560))?

It looks like they are created but somehow only the tests (the ones created last) make it into the index.

**Comment by carbolymer** [comment]
* [x] TODO: test against https://github.com/haskell/cabal/pull/9821

**Review by Jimbo4350 — APPROVED** [review]
LGTM however I'm not seeing the haddocks action in the summary.

File: .github/workflows/github-page.yml [hunk 1]
```diff
@@ -7,134 +7,59 @@ on:
...
-        archive-uri: https://iohk.cache.haskellworks.io/${{ env.CABAL_CACHE_VERSION }}/${{ runner.os }}/${{ matrix.cabal }}/${{ matrix.ghc }}
-        skip: "${{ vars.CACHE_URI != '' }}"
-        enable-save: false
-
-    - name: Install dependencies
-      run: cabal build all
+    - name: Build documentation
+      run: |
+        nix develop .\#haddockShell --command \
+          cabal haddock-project --output=./haddocks --internal --foreign-libraries
```
Comment (coot) [inline .github/workflows/github-page.yml:46]: :smile: :tada:

**Pull request merged**